### PR TITLE
Add notes on the watch port feature

### PR DESCRIPTION
**Pull request by jafingerhut**

### CODE REVIEW

File: docs/v1/P4Runtime-Spec.mdk [hunk 1]
```diff
@@ -3439,7 +3439,13 @@ An `ActionProfileGroup` entity update message has the following fields:
       `INVALID_ARGUMENT` if the client attempts to use it.
     * `watch` is the controller-defined 32-bit port number that the member's
       liveness depends on. At runtime, the member must be excluded from
-      selection if the watch port is down.
+      selection if the watch port is down.  See Section
+      [#action-selector-constraints] for notes on the behavior if all members in
+      a group are excluded for this reason. If `watch` is 0, then the member is
+      always included in the selection, regardless of the status of the
+      ports. The value must be 0 or the port number of an existing port on the
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk:3446]: s/ports/port?

File: docs/v1/P4Runtime-Spec.mdk [hunk 2]
```diff
@@ -3439,7 +3439,13 @@ An `ActionProfileGroup` entity update message has the following fields:
...
+      device, otherwise the server must return `INVALID_ARGUMENT` if the client
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk:3447]: "if the client attempts to use any other value" sounds a bit redundant given the beginning of the sentence.

File: docs/v1/P4Runtime-Spec.mdk [hunk 3]
```diff
@@ -3688,6 +3697,16 @@ The PSA specification includes a discussion on how to implement
 `psa_empty_group_action` in software in the P4Runtime server
 [@PSAEmptyGroupActionAppendix].
 
+If a P4Runtime implementation does support `psa_empty_group_action`, that action
+should be executed when an action selector group is selected that uses the watch
+port feature, and every member of the group has a watched port that is down.
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk:3702]: s/watched/watch for consistency?

**Comment by antoninbas** [comment]
@jafingerhut one more thing, please append one line at the end of the revision history section for this change

File: docs/v1/P4Runtime-Spec.mdk (hunk 1 above)
Comment (jafingerhut) [reply docs/v1/P4Runtime-Spec.mdk:3446]: I tried a change that rewords the sentence a bit to this: "If `watch` is 0, then the member is always included in the selection, regardless of the status of any port of the device."

It sounds a bit strange to me to say "regardless of the status of the port", when there is no "the port" involved if watch is 0.

File: docs/v1/P4Runtime-Spec.mdk (hunk 2 above)
Comment (jafingerhut) [reply docs/v1/P4Runtime-Spec.mdk:3447]: removed that phrase, leaving the rest of the sentence as it was.

File: docs/v1/P4Runtime-Spec.mdk (hunk 3 above)
Comment (jafingerhut) [reply docs/v1/P4Runtime-Spec.mdk:3702]: changed

**Comment by jafingerhut** [comment]
@antoninbas Commit 4 on this PR adds a proposed item to the revision history mentioning this and my earlier action profile/selector/watch port changes.  Let me know if you want it to explain more than what I've added there -- I kept it very brief.

File: docs/v1/P4Runtime-Spec.mdk [hunk 4]
```diff
@@ -3439,7 +3439,12 @@ An `ActionProfileGroup` entity update message has the following fields:
       `INVALID_ARGUMENT` if the client attempts to use it.
     * `watch` is the controller-defined 32-bit port number that the member's
       liveness depends on. At runtime, the member must be excluded from
-      selection if the watch port is down.
+      selection if the watch port is down.  See Section
+      [#action-selector-constraints] for notes on the behavior if all members in
+      a group are excluded for this reason. If `watch` is 0, then the member is
+      always included in the selection, regardless of the status of any port of
+      the device.  The value must be 0 or the port number of an existing port on
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk:3446]: That may be redundant with "controller-defined 32-bit port number" above, but do you think we should emphasize here and use "The value must be 0 or the SDN port number"?

Comment (jafingerhut) [reply docs/v1/P4Runtime-Spec.mdk]: Added "SDN" qualifier to "port number"

**Comment by jafingerhut** [comment]
I am in no personal rush to have this merged.  I just wanted to comment that it seems that Github is in a "Merging is blocked" status because it seems not to have "heard" back from Travis that all is well.

**Comment by antoninbas** [comment]
@jafingerhut I updated the Github branch protection settings and it looks ok now. This is caused by the transition from travis-ci.org to travis-ci.com.

If you have another PR that is stuck, you can try to close and re-open it.

**Pull request merged**